torchsynth/module.py

```diff
@@ -586,6 +548,124 @@ def _forward(self) -> Signal:
+            for j in range(n_output):
+                self.parameter_ranges.append(
+                    ModuleParameterRange(
+                        0.0,
+                        1.0,
+                        curve=curves[i],
+                        name=f"level{i}_{j}",
+                        description=f"Mix level for input {i} to output {j}",
+                    )
+                )
+
+        super().__init__(synthglobals, **kwargs)
+        self.n_input = n_input
+        self.n_output = n_output
+
+    def forward(self, *signals: Signal) -> Tuple[Signal]:
+        """
+        Mix together a set of modulation signals.
+        """
+
+        # Get params into batch_size x n_output x n_input matrix
+        params = torch.stack([self.p(p) for p in self.torchparameters], dim=1)
+        params = params.view(self.batch_size, self.n_input, self.n_output)
+        params = torch.swapaxes(params, 1, 2)
+
+        # Make sure there is the same number of input signals as mix params
+        assert len(signals) == params.shape[2]
+        signals = torch.stack(signals, dim=1)
+
+        modulation = torch.chunk(torch.matmul(params, signals), self.n_output, dim=1)
+        return tuple(m.squeeze(1).as_subclass(Signal) for m in modulation)
+
+
+class AudioMixer(SynthModule):
+    """
+    Sums together a set of N signals together and applies normalization if required
+    """
+
+    def __init__(
+        self,
+        synthglobals: SynthGlobals,
+        n_input: int,
+        curves: Optional[List[float]] = None,
+        **kwargs: Dict[str, T],
+    ):
+        # Parameter curves can be used to modify the parameter mapping
+        # for each input modulation source to the outputs
+        if curves is not None:
+            assert len(curves) == n_input
+        else:
+            curves = [1.0] * n_input
+
+        # Need to create the parameter ranges before calling super().__init
+        self.parameter_ranges = []
+        for i in range(n_input):
+            self.parameter_ranges.append(
+                ModuleParameterRange(
+                    -80.0,
+                    0.0,
+                    curve=curves[i],
+                    name=f"level{i}",
+                    description=f"Mix level for input {i} id dB",
+                )
+            )
+
+        super().__init__(synthglobals, **kwargs)
+        self.n_input = n_input
+
+    def _forward(self, *signals: Signal) -> Signal:
+
+        # Turn params into matrix and convert from dB
+        params = torch.stack([self.p(p) for p in self.torchparameters], dim=1)
+        params = torch.pow(10.0, params / 20.0)
+
+        # Make sure we received the correct number of input signals
+        signals = torch.stack(signals, dim=1)
+        assert signals.shape[1] == params.shape[1]
+
+        # Mix signals and normalize output if required
+        output = torch.matmul(params.unsqueeze(1), signals).squeeze(1)
+        return util.normalize_if_clipping(output)
```

Comment:
do we just want to normalize all output? Without some sounds come out as essentially silence (you can't tell unless you turn off ipd.Audio normalization)

Reply:
@jorshi please create an issue for this, we will decide based upon MMD experiments

Reply:
#202